Cleanup polymorphic builtins

tautschnig opened the pull request:
The main aim of this PR was the complete support for all `__atomic_*` and `__sync_*` built-ins provided by GCC. While working on this I noticed redundancies and incompleteness in both C and C++ type checking for those. I have eliminated the redundant bits and moved all the processing out of goto-program conversion, because this is entirely C/C++-specific code. Please review commit-by-commit.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
De-duplicate type checking of GCC polymorphic built-ins

The C front-end handled a few of them, while the C++ front-end handled a lot
more. Implement it just once, in the C front-end (which the C++ front-end then
makes use of).

Also add various identifiers of the built-ins to the set of known irep
IDs.

tautschnig committed:
Factor out __sync_lock_release

Its signature is quite different, and we will do more precise type checking of
each of the __sync_* primitives, all of which require more arguments.

tautschnig committed:
Factor out __sync_*_compare_and_swap

They require at least three arguments, and thus require type checking different
from the other __sync_* primitives.

tautschnig committed:
Prepare __sync_* regression tests for functional testing

Cleanup the directory structure to gather all tests of a primitive in a single
directory. For now these are just the type checking tests, functional tests will
be added in a subsequent commit.

tautschnig committed:
Functional tests of __sync_{bool,val}_compare_and_swap and __sync*fetch*

This is to confirm that they (at least in a single-threaded context) work
correctly.

tautschnig committed:
C library: implement non-polymorphic __atomic_* built-ins

This avoids spurious verification results in case code uses them.

tautschnig committed:
GCC polymorphic built-ins: create a symbol per type

Each type requires a separate declaration and definition.

tautschnig committed:
The __sync*fetch* primitives require at least two arguments

Make that requirement explicit during type checking and also fix the second
parameter to be of the subtype of the first (pointer-typed) parameter.

tautschnig committed:
Set parameter names of GCC polymorphic built-ins

This is in preparation of actually implementing them as separate functions.

tautschnig committed:
Implement various __sync_* built-ins in the C front-end

These are entirely specific to the C/C++ front-end, there is no need to
process them as late as goto-program conversion. First, this avoids a need for
(partial) re-type-checking. Second, we can as next step support instantiations
with multiple types.

tautschnig committed:
Fix types of __atomic_*fetch* GCC built-ins

We had already checked that three arguments were provided, but then constructed
types that only required one parameter.

tautschnig committed:
Implement __sync_*fetch* primitives via __atomic_*fetch*

This is what GCC documentation suggests, and we thus have both the __atomic_*
and the __sync_* variants implemented.

smowton commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -3468,3 +3445,355 @@ void c_typecheck_baset::make_constant_index(exprt &expr)
+  // assumptions that would require type checking
+
+  if(
+    identifier == "__sync_fetch_and_add" ||
```
While you're here, consider making these `static irep_idt`s to save hashing the string every time this is called

tautschnig replied:
Done.

smowton commented on regression/cbmc/sync_X_and_fetch-1/main.c:
```diff
@@ -0,0 +1,16 @@
+
+int main()
+{
+#ifndef _WIN32
```
Consider use of a tag in the .desc file rather than an ifndef (it's clearer if on Win32 the test reads SKIPPED)

tautschnig replied:
Done.

smowton commented on src/ansi-c/library/gcc.c:
```diff
@@ -32,3 +32,111 @@ inline void __sync_synchronize(void)
+
+/* FUNCTION: __atomic_test_and_set */
+
+void __atomic_thread_fence(int memorder);
```
Prototyped this twice?

tautschnig replied:
Yes, intentionally: each of the `FUNCTION:` sections is parsed individually, and thus needs to have all the necessary prototypes.

smowton commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -1935,18 +1936,33 @@ void c_typecheck_baset::typecheck_side_effect_function_call(
           identifier, expr.arguments(), f_op.source_location()))
       {
-        if(!symbol_table.has_symbol(gcc_polymorphic->get_identifier()))
+        irep_idt identifier_w_type = gcc_polymorphic->get_identifier();
```
Make some tests for the name mangling? Are there any particular cases where you might foresee the mangling producing an invalid name?

tautschnig replied:
As those names are only for internal use we are rather free to use whatever strings are formed here. But certainly I should at least have a test exercising this, which I have now added.

smowton commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -1937,10 +1937,33 @@ void c_typecheck_baset::typecheck_side_effect_function_call(
+            parameters[i].set_base_name(new_symbol.base_name);
+
+            symbolt *symbol_ptr;
+            move_symbol(new_symbol, symbol_ptr);
```
Considering `symbol_ptr` is not used, just use `add`?

tautschnig replied:
Done.

smowton commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -3868,9 +3868,11 @@ optionalt<symbol_exprt> c_typecheck_baset::typecheck_gcc_polymorphic_builtin(
-    symbol_exprt result(identifier, t);
+      {code_typet::parametert(ptr_arg.type()),
+       code_typet::parametert(ptr_arg.type().subtype()),
+       code_typet::parametert(signed_int_type())},
```
This change didn't resolve a test? Add one?

tautschnig replied:
Done.

smowton reviewed (commented):
The first few commits enable some previously-failing tests, but later ones seem to make larger changes without adding any new ones or enabling any further failing ones

peterschrammel commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -3468,3 +3445,355 @@ void c_typecheck_baset::make_constant_index(exprt &expr)
+    // adjust return type of function to match pointer subtype
+    if(arguments.empty())
+    {
+      error().source_location = source_location;
```
consider using proper exception

tautschnig replied:
Yes, but not in this PR I'd say. We need to fix this all over the C/C++ front-end.

peterschrammel commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -3468,3 +3445,355 @@ void c_typecheck_baset::make_constant_index(exprt &expr)
+    code_typet t{{code_typet::parametert(ptr_arg.type())}, return_type};
+    t.make_ellipsis();
+    symbol_exprt result{identifier, t};
+    result.add_source_location() = source_location;
```
intuitive API... :see_no_evil:

tautschnig replied:
Are you suggesting that `symbol_exprt` should also have a constructor that takes a `source_locationt`?

peterschrammel replied:
Not necessarily, but that is a side discussion. Nothing for this PR.
I'd expect either `result.add_source_location(source_location)` (If it is something to be "added") or
probably `result.set_source_location(source_location)` or
maybe `result.source_location() = source_location`.

peterschrammel commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -3468,3 +3445,355 @@ void c_typecheck_baset::make_constant_index(exprt &expr)
   }
 }
+
+optionalt<symbol_exprt> c_typecheck_baset::typecheck_gcc_polymorphic_builtin(
```
Consider moving this mega-function into a separate file and splitting into smaller ones

tautschnig replied:
Move to a separate file done, splitting up is still to-be-done.

tautschnig replied:
Refactoring completed.

peterschrammel commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -3881,3 +3888,349 @@ optionalt<symbol_exprt> c_typecheck_baset::typecheck_gcc_polymorphic_builtin(
   return {};
 }
+
+static symbolt result_symbol(
```
long method

tautschnig replied:
Done.

tautschnig commented:
> The first few commits enable some previously-failing tests, but later ones seem to make larger changes without adding any new ones or enabling any further failing ones

Tests added.

tautschnig committed:
Refactor typechecking of GCC polymorphic built-ins

No functional changes, just moving code to a file of its own and
breaking large functions into smaller ones.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 6b29574db68721cda7c3adc8444cf504c197b283).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/105646273

smowton commented on regression/cbmc/CMakeLists.txt:
```diff
@@ -1,20 +1,45 @@
 )
+endif()
 
+if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "MSVC")
```
use `set(...)` and `string(CONCAT ...)` to avoid duplication, similar to Makefile case

tautschnig replied:
Thanks!

smowton commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -1952,18 +1953,33 @@ void c_typecheck_baset::typecheck_side_effect_function_call(
           identifier, expr.arguments(), f_op.source_location()))
       {
-        if(!symbol_table.has_symbol(gcc_polymorphic->get_identifier()))
+        irep_idt identifier_w_type = gcc_polymorphic->get_identifier();
```
Expand `_w_` here and below? Is that `_write_` or something else?

tautschnig replied:
It's "with" - I'll spell that out :-)

smowton commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -3897,3 +3904,349 @@ optionalt<symbol_exprt> c_typecheck_baset::typecheck_gcc_polymorphic_builtin(
+  symbol_tablet &symbol_table)
+{
+  symbolt symbol;
+  symbol.name = id2string(identifier) + "::1::result";
```
Always pretending to belong to source line 1?

tautschnig replied:
No, the first block/scope (which truly is the case, although the exact number wouldn't matter much either as long as it is unique).

smowton reviewed (approved):
A few comments, haven't reviewed this thoroughly due to size though

tautschnig committed:
Implement __atomic_{load,store,exchange,compare_exchange}(_n)

This completes the list of GCC atomic built-ins. Also use
__atomic_compare_exchange to implement __sync_bool_compare_and_swap.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: d33bd83992c98cf15d6d3f6ed18f4c797f27fd3b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/105804853

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 02bdd948d154152b84d08129dff7c66e986affe9).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/112101315

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 11417aaf6de3419d8ab05ea23565e97049a72dbd).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/112929258

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 471f85cb2a80a24a8312fcafd6ff8c29c81caf22).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/113742938
